src/hashing.rs

```diff
@@ -0,0 +1,437 @@
+//! - Message authentication codes: these provide a keyed output against a message
+//! - Group generators: these produce publicly-verifiable curve group elements
+//! - Other protocol-specific hash functions with fixed input and output requirements
+//!
+//! ## Key derivation functions
+//!
+//!
+//!
+//! ## Message authentication codes
+//!
+//! From [hmac]: Simply calculating `hash(key + msg)` to obtain a MAC (message authentication code) is generally
+//! insecure because of a length-extension attack. This is because most hashing algorithms allow you to extend the
+//! message without knowing the key to obtain another valid MAC. For this reason, the HMAC specification prescribes
+//! using `hash(key'|hash(key*|msg))`.
+//!
+//! By using this API to generate an `Hmac` type, you don't need to worry about these details. The specification is
+//! taken care of by the API.
+//!
+//! The use of marker traits also ensures that you won't accidentally use hashing algorithms that are unsuitable for
+//! generating message authentication codes.
+//!
+//! The HMAC specification is described in [IETF RFC 2104](https://datatracker.ietf.org/doc/html/rfc2104).
+//!
+//! ## Group generators
+//!
+//! While this can also be used for Fiat-Shamir transcripts, consider Merlin instead.
+//!
+//!
+//! [hmac]: https://en.wikipedia.org/wiki/HMAC#Design_principles "HMAC: Design principles"
+
+use std::collections::HashMap;
+
+use digest::{Digest, Output};
+
+use crate::{errors::HashingError, keys::SecretKey, ristretto::RistrettoSecretKey};
+
+/// A utility trait, provider semi-hardcoded metadata about a given hashing utility function. Its primary purpose is to
+/// provide the domain separation tag; but you're also able to extract some data from implementations, including the
+/// version and category tag.
+///
+/// You should seldom, if ever, need to create instances of structs that implement this trait. They are designed to be
+/// handled at compile-time via the type system.
+///
+/// For example,
+/// ```edition2018
+/// use tari_crypto::hashing::{HmacV1, SafeHashMetadata};
+/// fn foo<M: SafeHashMetadata>() -> String {
+///     format!("{}/v{}", M::tag(), M::version())
+/// }
+///
+/// assert_eq!(foo::<HmacV1>().as_str(), "tari_hmac/v1");
+/// ```
+pub trait SafeHashMetadata {
+    /// Returns the version number for the metadata tag
+    fn version() -> u8;
+    /// Returns the category label for the metadata tag. For example, `tari_hmac`
+    fn tag() -> &'static str;
+    /// The domain separation tag is defined as `H(version | tag | output_size)`, where the version and tag are
+    /// typically hard-coded into the implementing type, and the output size is the length of the hash generated by
+    /// the hashing algorithm `D`.
+    fn hash_domain_separation_tag<D: Digest>(mut hasher: D) -> D {
+        hasher.update(&[Self::version()]);
+        encode_with_size(Self::tag(), hasher).chain(D::output_size().to_le_bytes())
```

Comment:
There may have been some misunderstanding around the intent for embedding an output size. The "sampling process" for a hash function should include the output size _required by the use case_, which may differ from the output size of the underlying digest. For example, a `SipHash` key is 32 bytes, but an underlying digest might output 64 bytes. Embedding the expected output size ensures that an implementer can't accidentally reuse a label but specify a different output size, which could produce keys with a subset relationship.

Reply:
Now you actually specify the type you want as the result of the key derivation, so the code can do additional checks.